File: implementations/go/bicolor/bicolorable.go
package main

import "fmt"


type  Color string
const (
	RED Color = "Red"
	GREEN Color = "Green"
)

func isBiColorable(n int, graph map[int][]int) bool {
	color := make(map[int]Color)
	for i := 0; i < n; i++ {
		queue := []int{i}
		color[i] = RED

		for len(queue) > 0{
			node := queue[0]
			queue := queue[1:]

			for _, neighbor := range(graph[node]){
				if _, ok := color[neighbor]; !ok{
					if color[node] == RED {
						color[neighbor] = GREEN
					} else {
						color[neighbor] = RED
					}
					queue = append(queue, neighbor)
				} else if color[node] == color[neighbor] {
					return false
				}
			}
		}
	}
	return true
}

func main(){
	graph := map[int][]int{
		0: {1, 3},
		1: {0, 2},
		2: {1, 3},
		3: {0, 2},
	}

	fmt.Println(isBiColorable(4, graph)) // true

	graph2 := map[int][]int{
		0: {1, 2},
		1: {0, 2},
		2: {0, 1},
	}

	fmt.Println(isBiColorable(3, graph2)) // false
}

File: implementations/javascript/biColorable.js
const Color = Object.freeze({
    RED: "red",
    BLUE: "blue"
});


function isBipartile(graph) {
    const color = new Map()

    for(const start of graph.keys()){
        if(!color.has(start)){
            const queue = [start]
            color.set(start, color.RED)
            while (queue.length > 0) {
                const node = queue.shift();

                for(const neigh of graph.get(node) || []) {
                    if(!color.has(neigh)) {
                        color.set(neigh, color.get(node) == Color.RED? Color.BLUE : Color.RED)
                        queue.push(neigh)
                    } else if (color.get(neigh) === color.get(node)) {
                        return false;
                    }
                }
            }
        }
    }
    return true;
}

const graph = new Map();
graph.set(1, [2, 3]);
graph.set(2, [1, 4]);
graph.set(3, [1, 4]);
graph.set(4, [2, 3]);

console.log(isBipartile(graph));